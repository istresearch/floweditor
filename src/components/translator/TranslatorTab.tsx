import { default as bindCallbacks } from 'auto-bind/react';
import classNames from 'classnames/bind';
import React from 'react';
import styles from './TranslatorTab.module.scss';
import i18n from 'config/i18n';
import { PopTab } from 'components/poptab/PopTab';
import { AssetMap, Asset, RenderNodeMap } from 'store/flowContext';
import { PopTabType, Type } from 'config/interfaces';
import { Action, Category, Case } from 'flowTypes';
import { getTypeConfig } from 'config';
import {
  findTranslations,
  getMergedByType,
  TranslationState,
  getFriendlyAttribute,
  getBundleKey
} from './helpers';
import CheckboxElement from 'components/form/checkbox/CheckboxElement';
import { UpdateTranslationFilters } from 'store/thunks';
import { getSwitchRouter } from 'components/flow/routers/helpers';
import { getType } from 'config/typeConfigs';
import { fakePropType } from 'config/ConfigProvider';

const cx: any = classNames.bind(styles);

export enum TranslationType {
  PROPERTY = 'property',
  CATEGORY = 'category',
  CASE = 'case'
}

export interface TranslationBundle {
  node_uuid: string;
  action_uuid?: string;
  translations: Translation[];
  translated: number;
  typeConfig: Type;
}

export interface Translation {
  type: TranslationType;
  attribute: string;
  from: string;
  to: string;
}

export interface TranslatorTabProps {
  localization: { [uuid: string]: any };
  language: Asset;
  languages: AssetMap;

  translationFilters: { categories: boolean; rules: boolean };

  nodes: RenderNodeMap;
  onToggled: (visible: boolean, tab: PopTabType) => void;
  onTranslationClicked: (bundle: TranslationBundle) => void;
  onTranslationOpened: (bundle: TranslationBundle) => void;
  onTranslationFilterChanged: UpdateTranslationFilters;
  popped: string;
}

export interface TranslatorTabState {
  visible: boolean;
  selectedTranslation: Translation;
  translationBundles: TranslationBundle[];
  optionsVisible: boolean;
  pctComplete: number;
  translationFilters: { categories: boolean; rules: boolean };
}

export class TranslatorTab extends React.Component<TranslatorTabProps, TranslatorTabState> {
  public static contextTypes = {
    config: fakePropType
  };

  constructor(props: TranslatorTabProps, context: any) {
    super(props);

    this.state = {
      visible: false,
      selectedTranslation: null,
      translationBundles: [],
      optionsVisible: false,
      pctComplete: 0,
      translationFilters: props.translationFilters || { categories: true, rules: true }
    };

    bindCallbacks(this, {
      include: [/^handle/, /^render/, /^toggle/]
    });
  }

  public componentWillUnmount(): void {
    if (this.state.visible) {
      this.handleTabClicked();
    }
  }

  public componentDidUpdate(prevProps: TranslatorTabProps, prevState: TranslatorTabState): void {
    // traceUpdate(this, prevProps, prevState);
    if (
      prevProps.translationFilters !== this.props.translationFilters ||
      prevProps.localization !== this.props.localization ||
      prevProps.language !== this.props.language ||
      !prevState.visible ||
      prevState.translationFilters !== this.state.translationFilters
    ) {
      this.handleUpdateTranslations();
    }
  }

  private handleUpdateTranslations(): void {
    const translationBundles: TranslationBundle[] = [];
    Object.keys(this.props.nodes).forEach((node_uuid: string) => {
      const renderNode = this.props.nodes[node_uuid];

      // check for router level translations
      if (
        renderNode.node.router &&
        (this.state.translationFilters.categories || this.state.translationFilters.rules)
      ) {
        const typeConfig = getTypeConfig(getType(renderNode));

        let translations: Translation[] = [];
        if (this.state.translationFilters.categories) {
          const localizeableKeys = ['name'];
          renderNode.node.router.categories.forEach((category: Category) => {
            translations.push(
              ...findTranslations(
                TranslationType.CATEGORY,
                localizeableKeys,
                category,
                this.props.localization
              )
            );
          });
        }

        if (this.state.translationFilters.rules) {
          const localizeableKeys = ['arguments'];
          const switchRouter = getSwitchRouter(renderNode.node);
          if (switchRouter) {
            switchRouter.cases.forEach((kase: Case) => {
              translations.push(
                ...findTranslations(
                  TranslationType.CASE,
                  localizeableKeys,
                  kase,
                  this.props.localization
                )
              );
            });
          }
        }

        if (translations.length > 0) {
          translationBundles.push({
            typeConfig,
            node_uuid,
            translations,
            translated: translations.filter((translation: Translation) => !!translation.to).length
          });
        }
      } else {
        // find attributes from each action
        renderNode.node.actions.forEach((action: Action) => {
          const typeConfig = getTypeConfig(action.type);
          const translations = findTranslations(
            TranslationType.PROPERTY,
            typeConfig.localizeableKeys || [],
            action,
            this.props.localization
          );

          if (translations.length > 0) {
            translationBundles.push({
              typeConfig,
              node_uuid,
              action_uuid: action.uuid,
              translations,
              translated: translations.filter((translation: Translation) => !!translation.to).length
            });
          }
        });
      }
    });

    const counts = { total: 0, complete: 0 };
    translationBundles.reduce((counts, bundle) => {
      counts.total += bundle.translations.length;
      counts.complete += bundle.translated;
      return counts;
    }, counts);

    const pctComplete = counts.total > 0 ? Math.round((counts.complete / counts.total) * 100) : 0;

    const bundles = translationBundles
      .filter((bundle: TranslationBundle) => bundle.translated < bundle.translations.length)
      .sort((a: TranslationBundle, b: TranslationBundle) => {
        return b.translations.length - b.translated - (a.translations.length - a.translated);
      });

    if (
      pctComplete !== this.state.pctComplete ||
      bundles.length !== this.state.translationBundles.length
    ) {
      this.setState({
        pctComplete,
        translationBundles: bundles
      });
    }
  }

  public handleTabClicked(): void {
    this.props.onToggled(!this.state.visible, PopTabType.TRANSLATOR_TAB);
    this.setState((prevState: TranslatorTabState) => {
      return { visible: !prevState.visible };
    });
  }

  public toggleOptions(): void {
    this.setState({ optionsVisible: !this.state.optionsVisible });
  }

  private toggleCategories(categories: boolean): void {
    this.setState(
      { translationFilters: { categories, rules: this.state.translationFilters.rules } },
      () => {
        this.props.onTranslationFilterChanged(this.state.translationFilters);
      }
    );
  }
  private toggleRules(rules: boolean): void {
    this.setState(
      { translationFilters: { rules, categories: this.state.translationFilters.categories } },
      () => {
        this.props.onTranslationFilterChanged(this.state.translationFilters);
      }
    );
  }

  private renderMissing(key: string, from: string, summary: string) {
    if (from) {
      return (
        <div key={this.props.language.id + key} className={styles.item}>
          <div className={styles.text + ' ' + styles.from_text}>{from}</div>
          <div className={styles.text + ' ' + styles.attribute}>{summary}</div>
        </div>
      );
    }
    return null;
  }

  private handleTranslationClicked(bundle: TranslationBundle) {
    this.props.onTranslationClicked(bundle);

    window.setTimeout(() => {
      this.props.onTranslationOpened(bundle);
    }, 750);
  }

  private handleChangeLanguageClick(e: any): void {
    this.context.config.onChangeLanguage(this.props.language.id, this.props.language.name);
    e.preventDefault();
    e.stopPropagation();
  }

  public render(): JSX.Element {
    // only show change button if container set up to handle it
    const showChangeButton = this.state.optionsVisible && this.context.config.onChangeLanguage;

    const classes = cx({
      [styles.visible]: this.state.visible,
      [styles.hidden]: this.props.popped && this.props.popped !== PopTabType.TRANSLATOR_TAB
    });

    const optionsClasses = cx({
      [styles.options]: true,
      [styles.options_visible]: this.state.optionsVisible,
      [styles.change_visible]: showChangeButton
    });

    const filledClasses = cx({
      [styles.filled]: true,
      [styles.hundredpct]: this.state.pctComplete === 100
    });

    const wrapperClasses = cx({
      [styles.translations_wrapper]: true,
      [styles.complete]: this.state.translationBundles.length === 0
    });

    return (
      <div className={classes}>
        <PopTab
          header={`${this.props.language.name} ${i18n.t('translation.label', 'Translations')}`}
          label={i18n.t('translation.header', 'Flow Translation')}
          color="#777"
          icon="fe-language"
          top="120px"
          popTop="100px"
          visible={this.state.visible}
          onShow={this.handleTabClicked}
          onHide={this.handleTabClicked}
          custom={false}
        >
          <div key={'translation_wrapper'} className={wrapperClasses}>
            {this.state.translationBundles.map((bundle: TranslationBundle) => {
              return (
                <div
                  key={this.props.language.id + getBundleKey(bundle)}
                  className={styles.translate_block}
                  onClick={() => {
                    this.handleTranslationClicked(bundle);
                  }}
                >
                  <div className={styles.needs_translation}>
                    <div className={styles.type_name}>{bundle.typeConfig.name}</div>
                    {this.renderMissing(
                      getBundleKey(bundle) + 'categories',
                      getMergedByType(bundle, TranslationState.MISSING, TranslationType.CATEGORY),
                      getFriendlyAttribute('categories')
                    )}

                    {this.renderMissing(
                      getBundleKey(bundle) + 'rules',

                      getMergedByType(bundle, TranslationState.MISSING, TranslationType.CASE),
                      getFriendlyAttribute('cases')
                    )}
                    {bundle.translations
                      .filter(
                        translation =>
                          !translation.to && translation.type === TranslationType.PROPERTY
                      )
                      .map(translation =>
                        this.renderMissing(
                          getBundleKey(bundle) + translation.from,
                          translation.from,
                          getFriendlyAttribute(translation.attribute)
                        )
                      )}
                  </div>
                </div>
              );
            })}
          </div>

          <div className={optionsClasses} onClick={this.toggleOptions}>
            <div className={styles.header}>
              <div className={styles.progress_bar}>
                <div
                  style={{ width: `${this.state.pctComplete}%` }}
                  className={filledClasses}
                ></div>
              </div>
              <div className={styles.toggle}>
                <span className="fe-arrow-up"></span>
              </div>
            </div>
            <div className={styles.filters}>
              <div className={styles.checkboxes}>
                <div>
                  <CheckboxElement
                    name={i18n.t('forms.categories', 'categories')}
                    checked={this.state.translationFilters.categories}
                    description={'Categories'}
                    checkboxClassName={styles.checkbox}
                    onChange={this.toggleCategories}
                  />
                </div>
                <div>
                  <CheckboxElement
                    name={i18n.t('forms.rules', 'rules')}
                    checked={this.state.translationFilters.rules}
                    description={'Rule Arguments'}
                    checkboxClassName={styles.checkbox}
                    onChange={this.toggleRules}
                  />
                </div>
              </div>
              <div className={styles.pct_complete}>{this.state.pctComplete}%</div>
            </div>
            <div className={styles.changeLanguage}>
              {showChangeButton && (
                <button onClick={this.handleChangeLanguageClick}>
                  {i18n.t('forms.use_as_default_language', 'Use as default language')}
                </button>
              )}
            </div>
          </div>
        </PopTab>
      </div>
    );
  }
}
