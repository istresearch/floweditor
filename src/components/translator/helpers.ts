import { desnake } from 'utils';
import i18next from 'i18next';
import { Translation, TranslationType, TranslationBundle } from './TranslatorTab';

export enum TranslationState {
  COMPLETE = 'complete',
  MISSING = 'missing'
}

export const getMergedByType = (
  bundle: TranslationBundle,
  state: TranslationState,
  type: TranslationType
) => {
  return bundle.translations
    .filter(translation => {
      switch (state) {
        case TranslationState.COMPLETE:
          return !!translation.to && translation.type === type;
        case TranslationState.MISSING:
          return !translation.to && translation.type === type;
      }
      return true;
    })
    .map(translation => (translation as any)[state === TranslationState.MISSING ? 'from' : 'to'])
    .join(', ');
};

export const findTranslations = (
  type: TranslationType,
  localizeableKeys: string[],
  localizable: any,
  localization: { [uuid: string]: any }
): Translation[] => {
  const translations: Translation[] = [];

  localizeableKeys.forEach((attribute: string) => {
    let keys = attribute.split('.');
    let from = localizable as any;
    let to: any = [];

    while (keys.length > 0 && from) {
      if (keys.length > 0 && from['uuid']) {
        to = (localization || {})[from['uuid']];
      }

      const path = keys.shift();
      if (to) {
        to = to[path];
      }
      from = from[path];
    }

    if (from) {
      if (to) {
        to = to.join(', ');
      }

      if (Array.isArray(from)) {
        from = from
          .map((obj: any) => {
            if (obj['name']) {
              return obj['name'];
            }
            if (obj['arguments']) {
              return obj['arguments'].join(' ');
            }
            return obj;
          })
          .join(', ');
      }

      if (from) {
        translations.push({
          type,
          attribute,
          from,
          to
        });
      }
    }
  });

  return translations;
};

export const getFriendlyAttribute = (attribute: string) => {
  if (attribute === 'categories') {
    return i18next.t('translation.attributes.categories', 'Categories');
  }

  if (attribute === 'cases') {
    return i18next.t('translation.attributes.cases', 'Rules');
  }

  if (attribute === 'text') {
    return i18next.t('translation.attributes.text', 'Message');
  }

  if (attribute === 'quick_replies') {
    return i18next.t('translation.attributes.quick_replies', 'Quick Replies');
  }

  if (attribute === 'templating.variables') {
    return i18next.t('translation.attributes.templates', 'Template Variables');
  }

  if (attribute === 'subject') {
    return i18next.t('translation.attributes.subject', 'Subject');
  }

  if (attribute === 'body') {
    return i18next.t('translation.attributes.body', 'Body');
  }

  return desnake(attribute);
};

export const getBundleKey = (bundle: TranslationBundle) => {
  return bundle.node_uuid + bundle.action_uuid;
};
