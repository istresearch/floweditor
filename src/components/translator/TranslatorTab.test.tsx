import * as React from 'react';
import { render } from 'test/utils';
import { TranslatorTab, TranslatorTabProps } from './TranslatorTab';
import {
  Spanish,
  createRenderNode,
  createSendMsgAction,
  createExit,
  createMatchRouter,
  createSendEmailAction
} from 'testUtils/assetCreators';
import { RenderNodeMap } from 'store/flowContext';
import { createUUID } from 'utils';
import { Category } from 'flowTypes';
import { getSwitchRouter } from 'components/flow/routers/helpers';

const translatorProps: TranslatorTabProps = {
  localization: {},
  nodes: {},
  languages: {},
  popped: null,

  // callbacks
  onToggled: jest.fn(),
  onTranslationClicked: jest.fn(),
  onTranslationOpened: jest.fn(),
  onTranslationFilterChanged: jest.fn(),

  translationFilters: {
    categories: true,
    rules: true
  },

  // translation
  language: Spanish
};

const createMessageNode = (
  message: string,
  quick_replies: string[] = [],
  variables: string[] = [],
  translation?: string
): { nodes: RenderNodeMap; localization: { [uuid: string]: any } } => {
  const nodes: RenderNodeMap = {};
  const localization: { [uuid: string]: any } = {};

  const sendMsg = createSendMsgAction({ text: message });
  sendMsg.quick_replies = quick_replies;

  if (variables.length > 0) {
    sendMsg.templating = {
      uuid: createUUID(),
      template: { uuid: createUUID(), name: 'My Template' },
      variables
    };
  }

  const renderNode = createRenderNode({
    actions: [sendMsg],
    exits: [createExit()]
  });

  nodes[renderNode.node.uuid] = renderNode;

  if (localization && translation) {
    localization[renderNode.node.actions[0].uuid] = { text: [translation] };
  }

  return { nodes, localization };
};

const createRouterNode = (
  categories: string[],
  translations?: string[]
): { nodes: RenderNodeMap; localization: { [uuid: string]: any } } => {
  const nodes: RenderNodeMap = {};
  const localization: { [uuid: string]: any } = {};

  const renderNode = createMatchRouter(categories);
  nodes[renderNode.node.uuid] = renderNode;

  if (localization && translations) {
    const router = getSwitchRouter(renderNode.node);
    router.categories.forEach((category: Category, idx: number) => {
      localization[category.uuid] = { name: [translations[idx]] };
    });
  }

  return { nodes, localization };
};

describe(TranslatorTab.name, () => {
  it('renders', () => {
    const { baseElement, getByText } = render(<TranslatorTab {...translatorProps} />);
    getByText('Spanish Translations');
    expect(baseElement).toMatchSnapshot();
  });

  it('finds message translations', () => {
    const { baseElement, getByText, rerender } = render(<TranslatorTab {...translatorProps} />);

    const updates = createMessageNode('Hello World!', ['yes', 'no'], ['var1', 'var2']);
    rerender(<TranslatorTab {...translatorProps} {...updates} />);

    // we pulled out all the localizable bits
    getByText('Hello World!');
    getByText('Quick Replies');
    getByText('Template Variables');

    getByText('0%');
    expect(baseElement).toMatchSnapshot();
  });

  it('finds email translations', () => {
    const { baseElement, getByText, rerender } = render(<TranslatorTab {...translatorProps} />);

    const renderNode = createRenderNode({
      actions: [createSendEmailAction({ subject: 'Urgent!', body: 'I need this yesterday' })],
      exits: [createExit()]
    });

    rerender(<TranslatorTab {...translatorProps} nodes={{ [renderNode.node.uuid]: renderNode }} />);

    // we pulled out all the localizable bits
    getByText('Subject');
    getByText('Body');

    getByText('0%');
    expect(baseElement).toMatchSnapshot();
  });

  it('finds split translations', () => {
    const { baseElement, getByText, rerender, queryByText } = render(
      <TranslatorTab {...translatorProps} />
    );

    let updates = createRouterNode(['Red', 'Green', 'Blue']);
    rerender(<TranslatorTab {...translatorProps} {...updates} />);

    // category list
    getByText('Red, Green, Blue, Other');

    // rule list
    getByText('red, green, blue');

    expect(baseElement).toMatchSnapshot();
  });

  it('handles partial category translation', () => {
    const { baseElement, getByText, rerender, queryByText } = render(
      <TranslatorTab {...translatorProps} />
    );

    // partial translation
    const updates = createRouterNode(['Red', 'Green', 'Blue'], ['Rojo', 'Verde']);
    rerender(<TranslatorTab {...translatorProps} {...updates} />);

    // category list
    getByText('Blue, Other');
    getByText('29%');

    expect(baseElement).toMatchSnapshot();
  });

  it('hides completed translations', () => {
    const { baseElement, getByText, rerender, queryByText } = render(
      <TranslatorTab {...translatorProps} />
    );

    const updates = createMessageNode('Hello World!', ['yes', 'no'], [], 'Hola Mundo!');

    rerender(<TranslatorTab {...translatorProps} {...updates} />);
    expect(queryByText('Hello World!')).toBeNull();
    getByText('50%');
    expect(baseElement).toMatchSnapshot();
  });
});
