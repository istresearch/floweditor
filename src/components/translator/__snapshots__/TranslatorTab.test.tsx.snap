// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TranslatorTab finds email translations 1`] = `
<body>
  <div>
    <div
      class=""
    >
      <div
        class="pop_wrapper pop_wrapper "
        style="right: -260px; top: 120px;"
      >
        <div
          class="tab_wrapper"
        >
          <div
            class="tab"
            style="background: rgb(119, 119, 119);"
          >
            <div
              class="icon"
            >
              <span
                class="fe-language"
              />
            </div>
            <div
              class="label"
            >
              Flow Translation
            </div>
          </div>
        </div>
        <div
          class="popped "
          style="border-color: #777; top: 100px; right: 0px;"
        >
          <div
            class="header"
            style="background: rgb(119, 119, 119);"
          >
            <div
              class="close fe-x"
            />
            <div
              class="header_label"
            >
              Spanish Translations
            </div>
          </div>
          <div
            class="body"
            style="background: rgb(119, 119, 119);"
          >
            <div
              class="translations_wrapper"
            >
              <div
                class="translate_block"
              >
                <div
                  class="needs_translation"
                >
                  <div
                    class="type_name"
                  >
                    Send Email
                  </div>
                  <div
                    class="item"
                  >
                    <div
                      class="text from_text"
                    >
                      Urgent!
                    </div>
                    <div
                      class="text attribute"
                    >
                      Subject
                    </div>
                  </div>
                  <div
                    class="item"
                  >
                    <div
                      class="text from_text"
                    >
                      I need this yesterday
                    </div>
                    <div
                      class="text attribute"
                    >
                      Body
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="options"
            >
              <div
                class="header"
              >
                <div
                  class="progress_bar"
                >
                  <div
                    class="filled"
                    style="width: 0%;"
                  />
                </div>
                <div
                  class="toggle"
                >
                  <span
                    class="fe-arrow-up"
                  />
                </div>
              </div>
              <div
                class="filters"
              >
                <div
                  class="checkboxes"
                >
                  <div>
                    <label
                      class="label"
                    >
                      <span
                        class="fe-check-square checkbox"
                        data-spec="checkbox"
                      />
                      <div
                        class="description_solo"
                        data-spec="description"
                      >
                        Categories
                      </div>
                    </label>
                  </div>
                  <div>
                    <label
                      class="label"
                    >
                      <span
                        class="fe-check-square checkbox"
                        data-spec="checkbox"
                      />
                      <div
                        class="description_solo"
                        data-spec="description"
                      >
                        Rule Arguments
                      </div>
                    </label>
                  </div>
                </div>
                <div
                  class="pct_complete"
                >
                  0
                  %
                </div>
              </div>
              <div
                class="changeLanguage"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`TranslatorTab finds message translations 1`] = `
<body>
  <div>
    <div
      class=""
    >
      <div
        class="pop_wrapper pop_wrapper "
        style="right: -260px; top: 120px;"
      >
        <div
          class="tab_wrapper"
        >
          <div
            class="tab"
            style="background: rgb(119, 119, 119);"
          >
            <div
              class="icon"
            >
              <span
                class="fe-language"
              />
            </div>
            <div
              class="label"
            >
              Flow Translation
            </div>
          </div>
        </div>
        <div
          class="popped "
          style="border-color: #777; top: 100px; right: 0px;"
        >
          <div
            class="header"
            style="background: rgb(119, 119, 119);"
          >
            <div
              class="close fe-x"
            />
            <div
              class="header_label"
            >
              Spanish Translations
            </div>
          </div>
          <div
            class="body"
            style="background: rgb(119, 119, 119);"
          >
            <div
              class="translations_wrapper"
            >
              <div
                class="translate_block"
              >
                <div
                  class="needs_translation"
                >
                  <div
                    class="type_name"
                  >
                    Send Message
                  </div>
                  <div
                    class="item"
                  >
                    <div
                      class="text from_text"
                    >
                      Hello World!
                    </div>
                    <div
                      class="text attribute"
                    >
                      Message
                    </div>
                  </div>
                  <div
                    class="item"
                  >
                    <div
                      class="text from_text"
                    >
                      yes, no
                    </div>
                    <div
                      class="text attribute"
                    >
                      Quick Replies
                    </div>
                  </div>
                  <div
                    class="item"
                  >
                    <div
                      class="text from_text"
                    >
                      var1, var2
                    </div>
                    <div
                      class="text attribute"
                    >
                      Template Variables
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="options"
            >
              <div
                class="header"
              >
                <div
                  class="progress_bar"
                >
                  <div
                    class="filled"
                    style="width: 0%;"
                  />
                </div>
                <div
                  class="toggle"
                >
                  <span
                    class="fe-arrow-up"
                  />
                </div>
              </div>
              <div
                class="filters"
              >
                <div
                  class="checkboxes"
                >
                  <div>
                    <label
                      class="label"
                    >
                      <span
                        class="fe-check-square checkbox"
                        data-spec="checkbox"
                      />
                      <div
                        class="description_solo"
                        data-spec="description"
                      >
                        Categories
                      </div>
                    </label>
                  </div>
                  <div>
                    <label
                      class="label"
                    >
                      <span
                        class="fe-check-square checkbox"
                        data-spec="checkbox"
                      />
                      <div
                        class="description_solo"
                        data-spec="description"
                      >
                        Rule Arguments
                      </div>
                    </label>
                  </div>
                </div>
                <div
                  class="pct_complete"
                >
                  0
                  %
                </div>
              </div>
              <div
                class="changeLanguage"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`TranslatorTab finds split translations 1`] = `
<body>
  <div>
    <div
      class=""
    >
      <div
        class="pop_wrapper pop_wrapper "
        style="right: -260px; top: 120px;"
      >
        <div
          class="tab_wrapper"
        >
          <div
            class="tab"
            style="background: rgb(119, 119, 119);"
          >
            <div
              class="icon"
            >
              <span
                class="fe-language"
              />
            </div>
            <div
              class="label"
            >
              Flow Translation
            </div>
          </div>
        </div>
        <div
          class="popped "
          style="border-color: #777; top: 100px; right: 0px;"
        >
          <div
            class="header"
            style="background: rgb(119, 119, 119);"
          >
            <div
              class="close fe-x"
            />
            <div
              class="header_label"
            >
              Spanish Translations
            </div>
          </div>
          <div
            class="body"
            style="background: rgb(119, 119, 119);"
          >
            <div
              class="translations_wrapper"
            >
              <div
                class="translate_block"
              >
                <div
                  class="needs_translation"
                >
                  <div
                    class="type_name"
                  >
                    Wait for Response
                  </div>
                  <div
                    class="item"
                  >
                    <div
                      class="text from_text"
                    >
                      Red, Green, Blue, Other
                    </div>
                    <div
                      class="text attribute"
                    >
                      Categories
                    </div>
                  </div>
                  <div
                    class="item"
                  >
                    <div
                      class="text from_text"
                    >
                      red, green, blue
                    </div>
                    <div
                      class="text attribute"
                    >
                      Rules
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="options"
            >
              <div
                class="header"
              >
                <div
                  class="progress_bar"
                >
                  <div
                    class="filled"
                    style="width: 0%;"
                  />
                </div>
                <div
                  class="toggle"
                >
                  <span
                    class="fe-arrow-up"
                  />
                </div>
              </div>
              <div
                class="filters"
              >
                <div
                  class="checkboxes"
                >
                  <div>
                    <label
                      class="label"
                    >
                      <span
                        class="fe-check-square checkbox"
                        data-spec="checkbox"
                      />
                      <div
                        class="description_solo"
                        data-spec="description"
                      >
                        Categories
                      </div>
                    </label>
                  </div>
                  <div>
                    <label
                      class="label"
                    >
                      <span
                        class="fe-check-square checkbox"
                        data-spec="checkbox"
                      />
                      <div
                        class="description_solo"
                        data-spec="description"
                      >
                        Rule Arguments
                      </div>
                    </label>
                  </div>
                </div>
                <div
                  class="pct_complete"
                >
                  0
                  %
                </div>
              </div>
              <div
                class="changeLanguage"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`TranslatorTab handles partial category translation 1`] = `
<body>
  <div>
    <div
      class=""
    >
      <div
        class="pop_wrapper pop_wrapper "
        style="right: -260px; top: 120px;"
      >
        <div
          class="tab_wrapper"
        >
          <div
            class="tab"
            style="background: rgb(119, 119, 119);"
          >
            <div
              class="icon"
            >
              <span
                class="fe-language"
              />
            </div>
            <div
              class="label"
            >
              Flow Translation
            </div>
          </div>
        </div>
        <div
          class="popped "
          style="border-color: #777; top: 100px; right: 0px;"
        >
          <div
            class="header"
            style="background: rgb(119, 119, 119);"
          >
            <div
              class="close fe-x"
            />
            <div
              class="header_label"
            >
              Spanish Translations
            </div>
          </div>
          <div
            class="body"
            style="background: rgb(119, 119, 119);"
          >
            <div
              class="translations_wrapper"
            >
              <div
                class="translate_block"
              >
                <div
                  class="needs_translation"
                >
                  <div
                    class="type_name"
                  >
                    Wait for Response
                  </div>
                  <div
                    class="item"
                  >
                    <div
                      class="text from_text"
                    >
                      Blue, Other
                    </div>
                    <div
                      class="text attribute"
                    >
                      Categories
                    </div>
                  </div>
                  <div
                    class="item"
                  >
                    <div
                      class="text from_text"
                    >
                      red, green, blue
                    </div>
                    <div
                      class="text attribute"
                    >
                      Rules
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="options"
            >
              <div
                class="header"
              >
                <div
                  class="progress_bar"
                >
                  <div
                    class="filled"
                    style="width: 29%;"
                  />
                </div>
                <div
                  class="toggle"
                >
                  <span
                    class="fe-arrow-up"
                  />
                </div>
              </div>
              <div
                class="filters"
              >
                <div
                  class="checkboxes"
                >
                  <div>
                    <label
                      class="label"
                    >
                      <span
                        class="fe-check-square checkbox"
                        data-spec="checkbox"
                      />
                      <div
                        class="description_solo"
                        data-spec="description"
                      >
                        Categories
                      </div>
                    </label>
                  </div>
                  <div>
                    <label
                      class="label"
                    >
                      <span
                        class="fe-check-square checkbox"
                        data-spec="checkbox"
                      />
                      <div
                        class="description_solo"
                        data-spec="description"
                      >
                        Rule Arguments
                      </div>
                    </label>
                  </div>
                </div>
                <div
                  class="pct_complete"
                >
                  29
                  %
                </div>
              </div>
              <div
                class="changeLanguage"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`TranslatorTab hides completed translations 1`] = `
<body>
  <div>
    <div
      class=""
    >
      <div
        class="pop_wrapper pop_wrapper "
        style="right: -260px; top: 120px;"
      >
        <div
          class="tab_wrapper"
        >
          <div
            class="tab"
            style="background: rgb(119, 119, 119);"
          >
            <div
              class="icon"
            >
              <span
                class="fe-language"
              />
            </div>
            <div
              class="label"
            >
              Flow Translation
            </div>
          </div>
        </div>
        <div
          class="popped "
          style="border-color: #777; top: 100px; right: 0px;"
        >
          <div
            class="header"
            style="background: rgb(119, 119, 119);"
          >
            <div
              class="close fe-x"
            />
            <div
              class="header_label"
            >
              Spanish Translations
            </div>
          </div>
          <div
            class="body"
            style="background: rgb(119, 119, 119);"
          >
            <div
              class="translations_wrapper"
            >
              <div
                class="translate_block"
              >
                <div
                  class="needs_translation"
                >
                  <div
                    class="type_name"
                  >
                    Send Message
                  </div>
                  <div
                    class="item"
                  >
                    <div
                      class="text from_text"
                    >
                      yes, no
                    </div>
                    <div
                      class="text attribute"
                    >
                      Quick Replies
                    </div>
                  </div>
                </div>
              </div>
            </div>
            <div
              class="options"
            >
              <div
                class="header"
              >
                <div
                  class="progress_bar"
                >
                  <div
                    class="filled"
                    style="width: 50%;"
                  />
                </div>
                <div
                  class="toggle"
                >
                  <span
                    class="fe-arrow-up"
                  />
                </div>
              </div>
              <div
                class="filters"
              >
                <div
                  class="checkboxes"
                >
                  <div>
                    <label
                      class="label"
                    >
                      <span
                        class="fe-check-square checkbox"
                        data-spec="checkbox"
                      />
                      <div
                        class="description_solo"
                        data-spec="description"
                      >
                        Categories
                      </div>
                    </label>
                  </div>
                  <div>
                    <label
                      class="label"
                    >
                      <span
                        class="fe-check-square checkbox"
                        data-spec="checkbox"
                      />
                      <div
                        class="description_solo"
                        data-spec="description"
                      >
                        Rule Arguments
                      </div>
                    </label>
                  </div>
                </div>
                <div
                  class="pct_complete"
                >
                  50
                  %
                </div>
              </div>
              <div
                class="changeLanguage"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`TranslatorTab renders 1`] = `
<body>
  <div>
    <div
      class=""
    >
      <div
        class="pop_wrapper pop_wrapper "
        style="right: -260px; top: 120px;"
      >
        <div
          class="tab_wrapper"
        >
          <div
            class="tab"
            style="background: rgb(119, 119, 119);"
          >
            <div
              class="icon"
            >
              <span
                class="fe-language"
              />
            </div>
            <div
              class="label"
            >
              Flow Translation
            </div>
          </div>
        </div>
        <div
          class="popped "
          style="border-color: #777; top: 100px; right: 0px;"
        >
          <div
            class="header"
            style="background: rgb(119, 119, 119);"
          >
            <div
              class="close fe-x"
            />
            <div
              class="header_label"
            >
              Spanish Translations
            </div>
          </div>
          <div
            class="body"
            style="background: rgb(119, 119, 119);"
          >
            <div
              class="translations_wrapper complete"
            />
            <div
              class="options"
            >
              <div
                class="header"
              >
                <div
                  class="progress_bar"
                >
                  <div
                    class="filled"
                    style="width: 0%;"
                  />
                </div>
                <div
                  class="toggle"
                >
                  <span
                    class="fe-arrow-up"
                  />
                </div>
              </div>
              <div
                class="filters"
              >
                <div
                  class="checkboxes"
                >
                  <div>
                    <label
                      class="label"
                    >
                      <span
                        class="fe-check-square checkbox"
                        data-spec="checkbox"
                      />
                      <div
                        class="description_solo"
                        data-spec="description"
                      >
                        Categories
                      </div>
                    </label>
                  </div>
                  <div>
                    <label
                      class="label"
                    >
                      <span
                        class="fe-check-square checkbox"
                        data-spec="checkbox"
                      />
                      <div
                        class="description_solo"
                        data-spec="description"
                      >
                        Rule Arguments
                      </div>
                    </label>
                  </div>
                </div>
                <div
                  class="pct_complete"
                >
                  0
                  %
                </div>
              </div>
              <div
                class="changeLanguage"
              />
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
